### PR TITLE
Add section on compiler flags

### PR DESCRIPTION
**Pull request by theotherjimmy**
@AnotherButler

### CODE REVIEW

**Comment by theotherjimmy** [comment]
Resolves https://github.com/ARMmbed/mbed-os/issues/3808

**Comment by theotherjimmy** [comment]
@RonEld Does this resolve the documentation issue in the aforementioned GitHub issue

**Comment by theotherjimmy** [comment]
Would this be better in `docs/cont/code_style.md` and linked in the porting guide?

File: docs/advanced/porting_guide.md [hunk 1]
```diff
@@ -36,6 +36,52 @@ Add the target description to ```mbed-os\targets\targets.json```:
...
+`-O3`      | Optimize for speed
+`-c99`     | C uses the C99 standard
+
+#### IAR EWARM
+
+Code submitted for a port must compile without warnings or errors using IAR EWARM with the following flags.
+
+flag             | meaning
+-----------------|---------------------------------------
+`-no_rtti`       | Disable run time type information (C++ only)
```
Comment (AnotherButler) [inline docs/advanced/porting_guide.md]: Please make "runtime" one word in all three tables.

File: docs/advanced/porting_guide.md [hunk 2]
```diff
@@ -36,6 +36,52 @@ Add the target description to ```mbed-os\targets\targets.json```:
...
+`-no_rtti` | Disable run time type information (C++ only)
+`-no_vla`  | Disable variable length arrays (C++ only)
+`-O0`      | Do not optimize
+`-O3`      | Optimize for speed
+`-c99`     | C uses the C99 standard
+
+#### IAR EWARM
+
+Code submitted for a port must compile without warnings or errors using IAR EWARM with the following flags.
+
```
Comment (AnotherButler) [inline docs/advanced/porting_guide.md]: This says that the errors use IAR EWARM and flags. Is that correct? If not, please change to "Code submitted for porting with IAR EWARM must compile without warnings or errors and must use the following flags" or something similar.

File: docs/advanced/porting_guide.md [hunk 3]
```diff
@@ -36,6 +36,52 @@ Add the target description to ```mbed-os\targets\targets.json```:
...
+`-fomit-frame-pointer`      | Do not keep a frame pointer
+`-fno-rtti`                 | Disable run time type information (C++ only)
+`-Os`                       | Optimize for size
+`-O0`                       | Do not optimize
+`-std=gnu99`                | C uses the GNU99 standard 
+`-std=gnu++98`              | C++ uses the GNU++98 standard
+
+#### ARM Compiler 5
+
+Code submitted as part of a port must compile without warnings or errors using ARM Compiler 5 with the following flags.
```
Comment (AnotherButler) [inline docs/advanced/porting_guide.md]: This says that the errors use ARM Compiler 5 and flags. Is that correct? If not, please change to "Code submitted for porting with ARM Compiler 5 must compile without warnings or errors and must use the following flags" or something similar.

File: docs/advanced/porting_guide.md [hunk 4]
```diff
@@ -36,6 +36,52 @@ Add the target description to ```mbed-os\targets\targets.json```:
 
 See the [mbed Target Documentation](mbed_targets.md) for more details on what this definition means.
 
+### Coding requirements
+
+All code submitted as part of a port must compile with GCC ARM Embedded, ARM Compiler 5 and IAR EWARM. Flags that port code must compile with for each compiler are listed below. The profiles used to build mbed OS are subsets of each of these flags with exactly one optimization setting.
+
+#### GCC ARM Embedded
+
+Code in a port must be compatible with the `softfp` float ABI and compile without warnings or errors using GCC ARM Embedded with the following flags.
```
Comment (AnotherButler) [inline docs/advanced/porting_guide.md]: This says that the errors use GCC ARM Embedded and flags. Is that correct? If not, please change to "Code in a port that uses GCC ARM Embedded must compile without warnings or errors and must use the following flags" or something similar.

File: docs/advanced/porting_guide.md [hunk 5]
```diff
@@ -36,6 +36,52 @@ Add the target description to ```mbed-os\targets\targets.json```:
 
 See the [mbed Target Documentation](mbed_targets.md) for more details on what this definition means.
 
+### Coding requirements
+
+All code submitted as part of a port must compile with GCC ARM Embedded, ARM Compiler 5 and IAR EWARM. Flags that port code must compile with for each compiler are listed below. The profiles used to build mbed OS are subsets of each of these flags with exactly one optimization setting.
```
Comment (AnotherButler) [inline docs/advanced/porting_guide.md]: For clarity, please change the second sentence to "Porting code must compile with specific flags for each compiler. See the list of these flags below." or something similar.

For clarity, please change the third sentence to "The profiles that build mbed OS are subsets of each of these flags. Each subset has one optimization setting." or something similar.

File: docs/advanced/porting_guide.md (hunk 4 above)
Comment (theotherjimmy) [inline docs/advanced/porting_guide.md]: That implies that ports are allowed to not use GCC ARM Embedded. That is not true.

File: docs/advanced/porting_guide.md (hunk 1 above)
Comment (theotherjimmy) [inline docs/advanced/porting_guide.md]: Ah spell check fighting me.

File: docs/advanced/porting_guide.md (hunk 5 above)
Comment (theotherjimmy) [inline docs/advanced/porting_guide.md]: Profiles do not build.

File: docs/advanced/porting_guide.md (hunk 4 above)
Comment (theotherjimmy) [inline docs/advanced/porting_guide.md]: Also, you dropped the `softfp` thing.

**Comment by RonEld** [comment]
@theotherjimmy This resolve the need to add requirement for `softfp`
Thanks

**Pull request merged**